pkg/server/plugin/upstreamauthority/spire/spire.go

```diff
@@ -89,121 +89,118 @@ func (m *Plugin) GetPluginInfo(context.Context, *plugin.GetPluginInfoRequest) (*
 }
 
 func (m *Plugin) SetLogger(log hclog.Logger) {
-	m.pollMtx.Lock()
-	defer m.pollMtx.Unlock()
-	m.log = log.Named("upstream-authority-spire")
+	m.log = log
 }
 
 func (m *Plugin) MintX509CA(request *upstreamauthority.MintX509CARequest, stream upstreamauthority.UpstreamAuthority_MintX509CAServer) error {
-	m.subscribeToPolling()
+	err := m.subscribeToPolling(stream.Context())
+	if err != nil {
+		return err
+	}
 	defer m.unsubscribeToPolling()
 
-	m.nodeMtx.RLock()
-	certChain, _, err := submitCSRUpstreamCA(stream.Context(), m.nodeClient, request.Csr)
-	m.nodeMtx.RUnlock()
+	certChain, err := m.submitCSRUpstreamCA(stream.Context(), request.Csr)
```

Comment:
This still seems racy. Consider the sequence of events:
1. Fetch loop grabs bundle with roots `A`
1. `submitCSRUpstreamCA` takes the bundle lock, gets a new CA and gets roots 'B', and updates the cached roots to `B`
1. Fetch loop finally gets the bundle lock and updates bundle with `A`
1. loop below picks up changes `A` and sends them down as a bundle. `B` is never sent.

The key takeaway is that only bundle content fetched _after_ the upstream RPC has finished minting the downstream CA or publishing the JWT key should be considered.

There are a few ways to implement this.

One approach is that there is a counter that is bumped whenever the X.509 roots are updated. The fetch loop could record the counter before it fetches. After it gets a new bundle, it could make sure the roots have not been updated out from underneath it by checking the counter. If the roots have been updated, then the roots in the bundle are ignored. A similar counter could be used for JWT keys.

Reply:
Oh, this scenario is being considered. Fetch loop fetches and updates the bundle within the bundle lock at: `fetchAndSetBundle`.

The same occurs with MintX509 in `submitCSRUpstreamCA` and with PublishJWTKey in: `pushAndSetInitialKeys`

Reply:
Ah, i missed that. TBH, I'm not in love with holding locks over RPC calls in general.

Reply:
Oh I see the point of that, I'll change it for the counter approach so we can remove that lock.